cpp/src/arrow/array/statistics.h

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <optional>
+#include <variant>
+
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+/// \brief Statistics for an Array
+///
+/// Apache Arrow format doesn't have statistics but data source such
+/// as Apache Parquet may have statistics. Statistics associate with
+/// data source can be read unified API via this class.
+struct ARROW_EXPORT ArrayStatistics {
+ public:
+  using ElementBufferType = std::variant<bool, int8_t, uint8_t, int16_t, uint16_t,
```

Comment:
Apart from var-len type, how do we support timestamp, internal and other primitive types?

Reply:
BTW, should we add an optional buffer to store values for string/binary types? Thus we can use std::string_view to represent them. If the min/max values are exact values, we can point them to the values in the arrow array and do not store any value in the optional buffer.

Reply:
> Apart from var-len type, how do we support timestamp, internal and other primitive types?

Ah, we need `arrow::DataType` for them. We can add it to `ArrayStatistics` or users can use it in associated `ArrayData`. The latter is a bit difficult to use but it may be preferred because it can reduce needed memory a bit.

> should we add an optional buffer to store values for string/binary types? Thus we can use std::string_view to represent them. If the min/max values are exact values, we can point them to the values in the arrow array and do not store any value in the optional buffer.

This is a discussion point in the PR description. In general, I think that `std::string_view` is better but there is a problem where we should store referred data as you mentioned. It may be costly to find the min/max values in the Arrow array because statistics provided by data source will not provide the value position.